[BEAM-797] A PipelineVisitor that creates a Spark-native pipeline.

aviemzur opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

aviemzur commented:
R: @amitsela

aviemzur commented:
Example of bounded pipeline output:

17/01/29 20:54:58 INFO BoundedDataset: Native Spark RDD pipeline:
(4) TextIO.Write/Write/Finalize.out MapPartitionsRDD[29] at mapPartitions at TransformTranslator.java:254 [Memory Deserialized 1x Replicated]
 |  TextIO.Write/Write/Initialize.out MapPartitionsRDD[14] at mapPartitions at TransformTranslator.java:254 [Memory Deserialized 1x Replicated]
 |  MapPartitionsRDD[13] at map at BoundedDataset.java:71 [Memory Deserialized 1x Replicated]
 |  ParallelCollectionRDD[12] at parallelize at BoundedDataset.java:71 [Memory Deserialized 1x Replicated]

aviemzur commented:
Example of unbounded pipeline output:

```
17/01/29 21:03:06 INFO UnboundedDataset: Native Spark DStream->RDD pipeline:
(8) MapPartitionsRDD[53] at mapPartitions at StreamingTransformTranslator.java:400 [Memory Serialized 1x Replicated]
 |  MapPartitionsRDD[52] at mapPartitions at StreamingTransformTranslator.java:400 [Memory Serialized 1x Replicated]
 |  MapPartitionsRDD[51] at flatMap at GroupCombineFunctions.java:86 [Memory Serialized 1x Replicated]
 |  MapPartitionsRDD[50] at map at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
 |  MapPartitionsRDD[49] at map at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
 |  MapPartitionsRDD[48] at mapToPair at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
 |  MapPartitionsRDD[47] at groupByKey at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
 |  ShuffledRDD[46] at groupByKey at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
 +-(8) MapPartitionsRDD[45] at mapToPair at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[44] at mapToPair at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[43] at map at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[42] at map at GroupCombineFunctions.java:73 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[41] at mapPartitions at StreamingTransformTranslator.java:400 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[40] at mapPartitions at StreamingTransformTranslator.java:400 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[39] at map at GroupCombineFunctions.java:286 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[38] at map at GroupCombineFunctions.java:286 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[37] at flatMapValues at GroupCombineFunctions.java:276 [Memory Serialized 1x Replicated]
    |  MapPartitionsRDD[36] at mapToPair at GroupCombineFunctions.java:276 [Memory Serialized 1x Replicated]
    |  ShuffledRDD[35] at combineByKey at GroupCombineFunctions.java:242 [Memory Serialized 1x Replicated]
    +-(8) MapPartitionsRDD[34] at mapToPair at GroupCombineFunctions.java:236 [Memory Serialized 1x Replicated]
       |  MapPartitionsRDD[33] at flatMapToPair at GroupCombineFunctions.java:219 [Memory Serialized 1x Replicated]
       |  MapPartitionsRDD[32] at mapPartitions at StreamingTransformTranslator.java:400 [Memory Serialized 1x Replicated]
       |  MapPartitionsRDD[31] at mapPartitions at StreamingTransformTranslator.java:400 [Memory Serialized 1x Replicated]
       |  MapPartitionsRDD[30] at map at StreamingTransformTranslator.java:203 [Memory Serialized 1x Replicated]
       |  UnionRDD[29] at window at StreamingTransformTranslator.java:193 [Memory Serialized 1x Replicated]
       |  MapPartitionsRDD[3] at map at UnboundedDataset.java:74 [Memory Serialized 1x Replicated]
       |  ParallelCollectionRDD[2] at parallelize at UnboundedDataset.java:74 [Memory Serialized 1x Replicated]
       |  MapPartitionsRDD[3] at map at UnboundedDataset.java:74 [Memory Serialized 1x Replicated]
       |  ParallelCollectionRDD[2] at parallelize at UnboundedDataset.java:74 [Memory Serialized 1x Replicated]
```

amitsela commented:
This could be useful as well, but I was going for something like:
If Beam Pipeline is: `Read.from(KafkaIO).apply(ParDo.of(new MyDoFn()))....` I'd want a `Visitor` that (instead of executing the pipeline) would simply write the actually applied Spark app as if it was written in Spark (as much as possible), so we get:
`KafkaUtils.createDirectStream(...).mapPartitions(new MyDoFn())...` - in case of other IOs it might be difficult since not all are supportted by Spark, but you could mock something. DoFn implementation details are clearly opaque too.
It's mostly a debug tool.

aviemzur commented:
@amitsela 

Sure.
I've committed the beginning of a second approach, more along the lines of what you are talking about.
It is not yet complete, but committed to show the approach.
WDYT?

Example output (Incomplete):
```
sparkContext.parallelize(Arrays.asList(???))
.mapPartitions(new org.apache.beam.runners.spark.examples.WordCount$ExtractWordsFn())
.mapPartitions(new org.apache.beam.sdk.transforms.MapElements$1())
.<combinePerKey>
.mapPartitions(new org.apache.beam.sdk.transforms.MapElements$1())
.sparkContext.parallelize(Arrays.asList(???))
.mapPartitions(new org.apache.beam.sdk.io.Write$Bound$1())
.<viewAsSingleton>
.<window>
.mapPartitions(new org.apache.beam.sdk.transforms.MapElements$1())
.groupByKey()
.mapPartitions(new org.apache.beam.sdk.io.Write$Bound$WriteShardedBundles())
.<viewAsIter>
.mapPartitions(new org.apache.beam.sdk.io.Write$Bound$2())
```

amitsela commented:
@aviemzur looking good. As for Write and its weird composite, I can explain it, but generally this is where things become tricky.. knowing when composites that are broken down into primitives to be executed by the Spark runner, would actually be a single transformation in Spark API (it's not directly translated because Beam has a different set of IO support, and a different approach and the runner should keep up with Beam).

aviemzur commented:
Added support for unbounded pipelines and tests and fixed what we talked about regarding composites (`TextIO.Write` in the previous comments).

What the output looks like right now:

Bounded pipeline:
```java
sparkContext.parallelize(Arrays.asList(...))
.mapPartitions(new org.apache.beam.runners.spark.examples.WordCount$ExtractWordsFn())
.mapPartitions(new org.apache.beam.sdk.transforms.Count$PerElement$1())
.<combinePerKey>
.mapPartitions(new org.apache.beam.runners.spark.examples.WordCount$FormatAsTextFn())
.<org.apache.beam.sdk.io.TextIO$Write$Bound>
```

Unbounded pipeline:
```java
KafkaUtils.createDirectStream(...)
.<window>
.mapPartitions(new org.apache.beam.runners.spark.SparkNativePipelineVisitorTest$FormatKVFn())
.mapPartitions(new org.apache.beam.sdk.transforms.Distinct$2())
.<combinePerKey>
.mapPartitions(new org.apache.beam.sdk.transforms.Keys$1())
.mapPartitions(new org.apache.beam.sdk.transforms.WithKeys$2())
.<org.apache.beam.sdk.io.kafka.AutoValue_KafkaIO_Write>
```

Still left to do:
* Return native translations for every translation still with a `// TODO: return meaningful Spark native operation` in `TransformTranslator` and `StreamingTransformTranslator`.
* Decide if the native operation strings should be returned from `TransformEvaluator`'s `toString()` method (current implementation) or add a more binding method `toNativeString()` to its interface to make sure that nothing is forgotten in future implementations of `TransformEvaluator` resulting in parts of the debug pipeline output looking like `.TransformTranslator$123@abc12`

aviemzur commented:
Rebased, added `toNativeTransform()` method to `TransformEvaluator` and translated all remaining translations to native.

@amitsela PTAL?

How it looks now:

Bounded Pipeline:
```java
sparkContext.parallelize(Arrays.asList(...))
_.mapPartitions(new org.apache.beam.runners.spark.examples.WordCount$ExtractWordsFn())
_.mapPartitions(new org.apache.beam.sdk.transforms.Count$PerElement$1())
_.combineByKey(..., new org.apache.beam.sdk.transforms.Combine$CombineFn$KeyIgnoringCombineFn(), ...)
_.groupByKey()
_.map(new org.apache.beam.sdk.transforms.Combine$CombineFn$KeyIgnoringCombineFn())
_.mapPartitions(new org.apache.beam.runners.spark.SparkNativePipelineVisitorTest$PlusOne())
sparkContext.union(...)
_.mapPartitions(new org.apache.beam.runners.spark.examples.WordCount$FormatAsTextFn())
_.<org.apache.beam.sdk.io.TextIO$Write$Bound>
```

Unbounded Pipeline:
```java
KafkaUtils.createDirectStream(...)
_.map(new org.apache.beam.sdk.transforms.windowing.FixedWindows())
_.mapPartitions(new org.apache.beam.runners.spark.SparkNativePipelineVisitorTest$FormatKVFn())
_.mapPartitions(new org.apache.beam.sdk.transforms.Distinct$2())
_.transform(new Function<JavaRDD<...>, JavaRDD<...>>() { @Override ... call(JavaRDD<...> rdd) { rdd.combineByKey(..., new org.apache.beam.sdk.transforms.Combine$CombineFn$KeyIgnoringCombineFn(), ...); } })
_.mapPartitions(new org.apache.beam.sdk.transforms.Keys$1())
_.mapPartitions(new org.apache.beam.sdk.transforms.WithKeys$2())
_.<org.apache.beam.sdk.io.kafka.AutoValue_KafkaIO_Write>
```

aviemzur commented:
Rebased on top of master.
@amitsela PTAL

amitsela commented:
@aviemzur test failures are unrelated, they are known flakes that I'm working to fix ASAP.
I'm afraid you'd have to rebase again though.. we're moving fast here 😉 
I will add that this looks great, but I have one comment about composites in general, but via the example of "bounded pipeline" (let's call it batch since we're trying to avoid this term, and since Spark calls it "batch") - `Count.PerElement` is shown as:
`mapPartitions` - makes sense as it maps into KV etc.
`combineByKey` - OK, it uses Spark's combine directly, 👍 
`groupByKey` - ? is it another call to GBK after Combine.PerKey ?
and followed by by another CombineFn call..
This is a bit unclear and I think a bit more thought needs to be put there. I can help with that!

aviemzur commented:
@amitsela 
The `groupByKey` and second `CombineFn` call are the translation of the next 2 transforms in the pipeline, not part of the `Count.perElement` transform translation.
```
...
PCollection<KV<String, Long>> wordCounts = lines
        .apply(new WordCount.CountWords());

    wordCounts
        .apply(GroupByKey.<String, Long>create())
        .apply(Combine.<String, Long, Long>groupedValues(Sum.ofLongs()));
...
```

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkNativePipelineVisitor.java:
```diff
@@ -0,0 +1,218 @@
+
+/**
+ * Spark-native pipeline - translates pipeline to a Spark native pipeline.
+ * Used for debugging purposes.
```
Add more docs maybe ? helps to understand how someone would contract a similar Spark-native pipeline...
Explain: not all translations are "native".... the idea is to help translate a Beam pipeline into a Spark one.
Give an example ?

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkNativePipelineVisitor.java:
```diff
@@ -0,0 +1,218 @@
+ * Spark-native pipeline - translates pipeline to a Spark native pipeline.
+ * Used for debugging purposes.
+ */
+public class SparkNativePipelineVisitor extends SparkRunner.Evaluator {
```
Is there a way to use the inheritance here ?
in `enterCompositeTransform` call `super` and if `return CompositeBehavior.ENTER_TRANSFORM` try to debug ?
in `doVisitTransform` `super` followed by debug again ?
You can make methods in `SparkRunner.Evaluator` protected if needed.

amitsela commented on runners/spark/pom.xml:
```diff
@@ -196,6 +196,11 @@
       <scope>provided</scope>
     </dependency>
     <dependency>
+      <groupId>commons-lang</groupId>
```
We really need this ?

aviemzur replied:
I needed it for `WordUtils#capitalize` in `SparkNativePipelineVisitor.NativeTransform#replaceFnString`

amitsela replied:
👍

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkNativePipelineVisitor.java:
```diff
@@ -0,0 +1,218 @@
+  }
+
+  public String getDebugString() {
+    return StringUtils.join(transforms, "\n");
```
Guava `Joiner` ? since it's already a dependency.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineOptions.java:
```diff
@@ -101,4 +101,8 @@ public String create(PipelineOptions options) {
   boolean getUsesProvidedSparkContext();
   void setUsesProvidedSparkContext(boolean value);
 
+  @Description("A special flag that prints translated Spark native pipeline instead of running it.")
```
Put this in `TestSparkPipelineOptions` instead.

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java:
```diff
@@ -223,9 +223,12 @@ public void run() {
-          pipeline.traverseTopologically(new Evaluator(new TransformTranslator.Translator(),
-                                                       evaluationContext));
-          evaluationContext.computeOutputs();
+          TransformTranslator.Translator translator = new TransformTranslator.Translator();
```
In `TestSparkRunner` ?

amitsela commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/UnboundedDataset.java:
```diff
@@ -75,4 +77,15 @@ public void action() {
+
+  @Override
+  public void printDebugString() {
+    dStream.foreachRDD(new Function<JavaRDD<WindowedValue<T>>, Void>() {
```
Why not simply LOG and call `action()` ?

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkNativePipelineVisitorTest.java:
```diff
@@ -0,0 +1,234 @@
+/**
+ * Test {@link SparkNativePipelineVisitor} with different pipelines.
+ */
+public class SparkNativePipelineVisitorTest {
```
change test names to ...batch/streamingPipeline instead of bounded/unbounded ?

aviemzur replied:
👍

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkNativePipelineVisitorTest.java:
```diff
@@ -0,0 +1,234 @@
+
+    String debugString = visitor.getDebugString();
+
+    System.out.println(debugString);
```
Remove printouts

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkNativePipelineVisitorTest.java:
```diff
@@ -0,0 +1,234 @@
+  public void setUp() throws IOException {
+    outputDir = tmpDir.newFolder("out");
+    outputDir.delete();
+    jsc = new JavaSparkContext("local[*]", "Existing_Context");
```
In tests, default is to reuse context, and there's a rule to force Reuse (yes/no) as well so I don't think you need to specifically relate to the context in `before`/`teardown`

aviemzur commented:
@amitsela rebased on top of master and addressed your comments.

aviemzur commented:
Retest this please.

aviemzur commented:
Run Spark RunnableOnService.

aviemzur commented:
Run Spark RunnableOnService.

aviemzur commented:
Run Spark RunnableOnService.

aviemzur commented:
Run Spark RunnableOnService

aviemzur commented:
Retest this please.

aviemzur commented:
Run Spark RunnableOnService

aviemzur commented:
@amitsela rebased on top of master and created `SparkRunnerDebugger` to be used instead of `TestSparkRunner` for this.

amitsela commented on runners/spark/src/test/java/org/apache/beam/runners/spark/SparkRunnerDebuggerTest.java:
```diff
@@ -0,0 +1,196 @@
+
+    wordCounts
+        .apply(MapElements.via(new WordCount.FormatAsTextFn()))
+        .apply(TextIO.Write.to(outputDir.getAbsolutePath()).withNumShards(3).withSuffix(".txt"));
```
This is not really executed, so no need for `TemporaryFolder` rule and `@Before` init, no ?

aviemzur commented:
@amitsela addressed your comment in my latest commit.
PTAL

amitsela commented:
LGTM!
